### PR TITLE
K8SPSMDB-284: Major Upgrade

### PR DESCRIPTION
**Pull request by blez**
[K8SPSMDB-284](https://jira.percona.com/browse/K8SPSMDB-284) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

This PR adds possibility to do major upgrade.
Downgrade will be done in separate PR

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scheduled-backup | passed |
| scaling | passed |
| operator-self-healing | passed |
| arbiter | passed |
| demand-backup-sharded | failed |
| security-context | passed |
| service-per-pod | failed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | failed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/6d08b635af342daee4cab1b0ab73e9654c61a573
image: `perconalab/percona-server-mongodb-operator:PR-625-6d08b635`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | failed |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | failed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/6d08b635af342daee4cab1b0ab73e9654c61a573
image: `perconalab/percona-server-mongodb-operator:PR-625-6d08b635`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| upgrade | passed |
| liveness | passed |
| upgrade-sharded | passed |
| smart-update | passed |
| version-service | failed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/0b8e3764f007c40c23e4a0391746f736746e0233
image: `perconalab/percona-server-mongodb-operator:PR-625-0b8e3764`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | failed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | failed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/792310e8319307a438aacf473f445fb833b389ed
image: `perconalab/percona-server-mongodb-operator:PR-625-792310e8`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | failed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/c49d1a688f9be27f1bb3b3837c09cb1de91d9618
image: `perconalab/percona-server-mongodb-operator:PR-625-c49d1a68`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| upgrade | passed |
| liveness | passed |
| upgrade-sharded | passed |
| smart-update | passed |
| version-service | failed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/178a913202c5c8891174146a76493346bc808e07
image: `perconalab/percona-server-mongodb-operator:PR-625-178a9132`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| security-context | passed |
| arbiter | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| upgrade-sharded | passed |
| version-service | failed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/8bc4cb8fc799aa38f7707ffba2526b8205f14df9
image: `perconalab/percona-server-mongodb-operator:PR-625-8bc4cb8f`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| operator-self-healing | passed |
| security-context | passed |
| service-per-pod | failed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | failed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/04cdbbb2bac3a74ae857b09e3f35f5183bc8d528
image: `perconalab/percona-server-mongodb-operator:PR-625-04cdbbb2`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | failed |
| rs-shard-migration | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| upgrade-sharded | passed |
| version-service | failed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/f912c2f98578ced33fb4de409b9c8f188a175960
image: `perconalab/percona-server-mongodb-operator:PR-625-f912c2f9`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| limits | passed |
| arbiter | passed |
| scaling | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| security-context | passed |
| scheduled-backup | passed |
| rs-shard-migration | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| data-sharded | passed |
| upgrade-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/f9a6535e808cab9960faf23303153d6da0d59821
image: `perconalab/percona-server-mongodb-operator:PR-625-f9a6535e`

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| limits | passed |
| demand-backup | passed |
| scaling | passed |
| scheduled-backup | passed |
| monitoring-2-0 | passed |
| security-context | passed |
| operator-self-healing | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| arbiter | passed |
| service-per-pod | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/197fa330899a5b5f8350cae00f751b40d182ea66
image: `perconalab/percona-server-mongodb-operator:PR-625-197fa330`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| demand-backup-sharded | passed |
| smart-update | passed |
| version-service | passed |
| upgrade | failed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/6075abb13afabda32423dcf345892c6832c898ef
image: `perconalab/percona-server-mongodb-operator:PR-625-6075abb1`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| storage | passed |
| users | passed |
| self-healing | passed |
| data-sharded | passed |
| operator-self-healing | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/6075abb13afabda32423dcf345892c6832c898ef
image: `perconalab/percona-server-mongodb-operator:PR-625-6075abb1`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| init-deploy | passed |
| smart-update | passed |
| limits | passed |
| version-service | passed |
| scaling | passed |
| users | passed |
| security-context | passed |
| data-sharded | passed |
| rs-shard-migration | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/6075abb13afabda32423dcf345892c6832c898ef
image: `perconalab/percona-server-mongodb-operator:PR-625-6075abb1`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| monitoring-2-0 | passed |
| limits | passed |
| demand-backup | passed |
| arbiter | passed |
| scaling | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| liveness | passed |
| upgrade | passed |
| smart-update | passed |
| version-service | passed |
| upgrade-sharded | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/2b30c8a452d0118f87d54abe610a28490758ec59
image: `perconalab/percona-server-mongodb-operator:PR-625-2b30c8a4`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| arbiter | passed |
| scheduled-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | failed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | passed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/5461f525bc7305a8bef56f941f9f0dfafb47bed6
image: `perconalab/percona-server-mongodb-operator:PR-625-5461f525`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | failed |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/5461f525bc7305a8bef56f941f9f0dfafb47bed6
image: `perconalab/percona-server-mongodb-operator:PR-625-5461f525`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| storage | passed |
| self-healing | passed |
| operator-self-healing | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/5461f525bc7305a8bef56f941f9f0dfafb47bed6
image: `perconalab/percona-server-mongodb-operator:PR-625-5461f525`

**Review by delgod — APPROVED** [review]

File: pkg/apis/psmdb/v1/psmdb_types.go [hunk 1]
```diff
@@ -122,9 +123,18 @@ func (us UpgradeStrategy) Lower() UpgradeStrategy {
 	return UpgradeStrategy(strings.ToLower(string(us)))
 }
 
+func OneOfUpgradeStrategy(a string) bool {
+	return strings.ToLower(a) == string(UpgradeStrategyLatest) ||
```
Comment (qJkee) [inline pkg/apis/psmdb/v1/psmdb_types.go:127]:
```
a = strings.ToLower(a)
return ......
```

File: pkg/apis/psmdb/v1/psmdb_types.go [hunk 2]
```diff
@@ -122,9 +123,18 @@ func (us UpgradeStrategy) Lower() UpgradeStrategy {
...
+		strings.ToLower(a) == string(UpgradeStrategyRecommended) ||
+		strings.ToLower(a) == string(UpgradeStrategyDiasbled) ||
+		strings.ToLower(a) == string(UpgradeStrategyNever)
```
Comment (qJkee) [inline pkg/apis/psmdb/v1/psmdb_types.go:130]: i think it's worth to accept `UpgradeStrategy` instead of casting all constants

Comment (blez) [reply pkg/apis/psmdb/v1/psmdb_types.go]: In any case we can pass any string here. So we should lowercse it which requires casting to string.
So I simplify it a bit, but accept string https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/a7211780084d45514a1eb75fdf8f0ca2f2bbf0fb

File: pkg/apis/psmdb/v1/psmdb_types.go (hunk 1 above)
Comment (blez) [reply pkg/apis/psmdb/v1/psmdb_types.go]: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/a7211780084d45514a1eb75fdf8f0ca2f2bbf0fb

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| init-deploy | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| arbiter | passed |
| scheduled-backup | passed |
| limits | passed |
| operator-self-healing | passed |
| service-per-pod | failed |
| scaling | passed |
| demand-backup-sharded | failed |
| security-context | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/a7211780084d45514a1eb75fdf8f0ca2f2bbf0fb
image: `perconalab/percona-server-mongodb-operator:PR-625-a7211780`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup | passed |
| limits | passed |
| scaling | passed |
| scheduled-backup | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| rs-shard-migration | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/625/commits/452b134bdf35ea04274f9dbf8beb5e42d62b3f6d
image: `perconalab/percona-server-mongodb-operator:PR-625-452b134b`

**Review by qJkee — APPROVED** [review]

**Pull request merged**